warehouse/query-core/src/main/java/datawave/query/jexl/JexlASTHelper.java

```diff
@@ -1195,43 +1256,116 @@ public static LiteralRange<BigDecimal> getBigDecimalBoundedRange(JexlNode[] chil
+     * Iterate through provided node's ancestors and return true if any are an instance of ASTOrNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.ASTOrNode
+     *
+     * @param node
+     * @return True if any ancestor is an instance of ASTOrNode. If not, then False.
+     */
+    public static boolean isDescendantOfOr(JexlNode node) {
         if (null != node && null != node.jjtGetParent()) {
             JexlNode parent = node.jjtGetParent();
             
             if (parent instanceof ASTOrNode) {
                 return true;
             }
             
-            return isWithinOr(parent);
+            return isDescendantOfOr(parent);
         }
         
         return false;
     }
     
-    public static boolean isWithinNot(JexlNode node) {
-        while (null != node && null != node.jjtGetParent()) {
+    /**
+     * Iterate through provided node's ancestors and return true if any are an instance of ASTNotNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.ASTNotNode
+     *
+     * @param node
+     * @return True if any ancestor is an instance of ASTNotNode. If not, then False.
+     */
+    public static boolean isDescendantOfNot(JexlNode node) {
+        if (null != node && null != node.jjtGetParent()) {
             JexlNode parent = node.jjtGetParent();
             
             if (parent instanceof ASTNotNode) {
                 return true;
             }
             
-            return isWithinNot(parent);
+            return isDescendantOfNot(parent);
         }
         
         return false;
     }
     
-    public static boolean isWithinAnd(JexlNode node) {
-        while (null != node && null != node.jjtGetParent()) {
+    /**
+     * Iterate through provided node's ancestors and return true if any are an instance of ASTAndNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.ASTAndNode
+     *
+     * @param node
+     * @return True if any ancestor is an instance of ASTAndNode. If not, then False.
+     */
+    public static boolean isDescendantOfAnd(JexlNode node) {
+        if (null != node && null != node.jjtGetParent()) {
             JexlNode parent = node.jjtGetParent();
             
             if (parent instanceof ASTAndNode) {
                 return true;
             }
             
-            return isWithinAnd(parent);
+            return isDescendantOfAnd(parent);
+        }
+        
+        return false;
+    }
+    
+    /**
+     * Iterate through provided node's ancestors and return true if any are an instance of the provided class. The provided class must extend JexlNode.
+     * 
+     * @see org.apache.commons.jexl2.parser.JexlNode
+     *
+     * @param node
+     * @param typeKey
+     * @return True if any ancestor is an instance of the provided class. If not, then False.
+     */
+    public static <T extends JexlNode> boolean isDescendantOfNodeType(JexlNode node, Class<T> typeKey) {
+        if (null != node && null != node.jjtGetParent()) {
```

Comment:
Instead of the recursive calls here, you could have a more memory efficient setup like this. I don't think you should make this change in this merge request, but maybe that's work doing as a separate unit of work. 
```
while (node.getParent() != null){
  node = node.getParent()
  if (node.isInstance(type)) {
    return true
  }
}
```

Reply:
Created https://github.com/NationalSecurityAgency/datawave/issues/1946 to address this